dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -57,81 +58,107 @@ void FineGrainedShuffleWriter<StreamWriterPtr>::finishWrite()
     }
 }
 
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::prepare(const Block & sample_block)
+{
+    header = sample_block.cloneEmpty();
+    num_columns = header.columns();
+    // fine_grained_shuffle_stream_count is in (0, 1024], and partition_num is uint16_t, so will not overflow.
+    num_bucket = partition_num * fine_grained_shuffle_stream_count;
+    ;
+    partition_key_containers_for_reuse.resize(collators.size());
+    resetScatterColumns();
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::flush()
+{
+    if (rows_in_blocks > 0)
+        batchWriteFineGrainedShuffle<false>();
+}
+
 template <class StreamWriterPtr>
 void FineGrainedShuffleWriter<StreamWriterPtr>::write(const Block & block)
 {
     RUNTIME_CHECK_MSG(
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
+
     size_t rows = block.rows();
     rows_in_blocks += rows;
     if (rows > 0)
     {
         blocks.push_back(block);
     }
 
-    if (static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size)
+    if (blocks.size() == fine_grained_shuffle_stream_count || static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size * fine_grained_shuffle_stream_count)
```

Comment:
It seems that `fine_grained_shuffle_batch_size * fine_grained_shuffle_stream_count` can be pre-calculated

Reply:
Yes, will change it.